docs/overview.md

```diff
@@ -0,0 +1,15 @@
+# IAVL Spec
+
+The IAVL tree is a versioned, snapshottable (immutable) AVL+ tree for persistent data.
+
+The purpose of this data structure is to provide persistent storage for key-value pairs (say to store account balances) such that a deterministic merkle root hash can be computed.  The tree is balanced using a variant of the [AVL algorithm](http://en.wikipedia.org/wiki/AVL_tree) so all operations are O(log(n)).
```

Comment:
To re-iterate what I said in the call. This is excellent documentation presenting the statusquo of the iavl tree. Maybe it is out of scope for this spec but we should definitely describe for instance why the tree is balancing (as opposed to for instance a sparse merkle tree where the merkle hash trail is always the same size)?
A document explaining the overall design space, the required/desired properties and explains the rationale for the current design (what does the iavl tree optimize for and why?) and advantages over other designs is needed. I'm happy to team up on this effort.

Reply:
Definitely agree that a rationale document is needed, don't believe its in scope for this PR tho. let's open an issue on it, and im happy to help out writing it up tho I could definitely benefit from more context/help on it

Reply:
That would be a huge help @Liamsi !!!! Happy to see you back in these parts 😉

Reply:
@Liamsi I'd be happy to pair on working on such a document. Its been something thats been on the back of my mind for awhile. (In particular because I think the I in IAVL is unnecessary, and that we can get better performance with a patricia trie variant that supports iteration)

Reply:
@AdityaSripal could you make some follow up issues then we can merge this

Reply:
ref #176